pennylane/transforms/core/transform_dispatcher.py

```diff
@@ -438,3 +477,32 @@ def is_informative(self):
     def final_transform(self):
         """``True`` if the transform needs to be executed"""
         return self._final_transform
+
+
+@functools.lru_cache
+def _create_transform_primitive(name):
+    try:
+        # pylint: disable=import-outside-toplevel
+        import jax
+    except ImportError:
+        return None
+
+    transform_prim = jax.core.Primitive(name)
+    transform_prim.multiple_results = True
+
+    @transform_prim.def_impl
+    def _(
+        *all_args, inner_jaxpr, args_slice, consts_slice, targs, tkwargs
+    ):  # pylint: disable=unused-argument
+        args = all_args[args_slice]
+        consts = all_args[consts_slice]
+
+        out = jax.core.eval_jaxpr(inner_jaxpr, consts, *args)
+        out = out if isinstance(out, Sequence) else [out]
+        return out
```

Comment:
But then where are we applying the transform?

Reply:
We're not. The intention right now is strictly to capture transforms. Though I should add that transforms do not get applied right now even without the changes in this PR with program capture enabled. The following is in `master`:

```python
import pennylane as qml
qml.capture.enable()

@qml.transforms.core.transform
def operation_z_to_hadamard(tape):
    """Transform that converts Z gates to H gates."""
    new_ops = [qml.H(wires=op.wires) if isinstance(op, qml.Z) else op for op in tape.operations]
    return [qml.tape.QuantumScript(new_ops, tape.measurements)], lambda res: res[0]


@qml.transforms.core.transform
def expval_z_obs_to_x_obs(tape):
    """Transform that converts Z observables for expectation values to X observables."""
    new_measurements = [
        (
            qml.expval(qml.X(mp.wires))
            if isinstance(mp, qml.measurements.ExpectationMP) and isinstance(mp.obs, qml.Z)
            else mp
        )
        for mp in tape.measurements
    ]
    return [qml.tape.QuantumScript(tape.operations, new_measurements)], lambda res: res[0]


@expval_z_obs_to_x_obs
@qml.qnode(qml.device("default.qubit", wires=3))
@operation_z_to_hadamard
def f():
    qml.Z(0)
    qml.X(0)
    return qml.expval(qml.Z(0))
```
```pycon
>>> f()
Array(-1., dtype=float64)
```

If only `operation_z_to_hadamard` or `expval_z_obs_to_x_obs` was applied, the output would be 0, and if both were applied, the output would be 1.

Reply:
Should we then `raise NotImplementedError` for the implementation isntead?

Reply:
We shouldn't pay too much attention to what master was doing with capture enabled, because the capture workflow was just roughly smashing basic behavior together. It wasn't worth our effort to add `NotImplementedError`'s for all the many, many things that are not implemented. That doesn't mean it was good behavior, just experimental behavior.

Now that we are paying attention to transforms, I think a `NotImplementedError` would be preferable. Otherwise, we might start assigning too much importance to this `def_impl`.

Reply:
Thanks, that makes sense to me. I'll replace with a `NotImplementedError` and update tests.